pipeline/cb-schema/json.ml

```diff
@@ -0,0 +1,122 @@
+type t =
+  (* Yojson.Safe.t = *)
+  [ `Null
+  | `Bool of bool
+  | `Int of int
+  | `Intlit of string
+  | `Float of float
+  | `String of string
+  | `Assoc of (string * t) list
+  | `List of t list
+  | `Tuple of t list
+  | `Variant of string * t option ]
+
+(* Yojson.Safe printing functions reproduced *)
+
+let pp_list sep ppx out l =
+  let pp_sep out () = Format.fprintf out "%s@ " sep in
+  Format.pp_print_list ~pp_sep ppx out l
+
+let is_atom : t -> bool = function
+  | `Null | `Bool _ | `Int _ | `Float _ | `String _ | `Intlit _
+  | `List []
+  | `Assoc []
+  | `Tuple []
+  | `Variant (_, None) ->
+      true
+  | `List _ | `Assoc _ | `Tuple _ | `Variant (_, Some _) -> false
+
+let is_atom_list l = List.for_all is_atom l
+
+(*
+     inside_box: indicates that we're already within a box that imposes
+     a certain style and we shouldn't create a new one. This is used for
+     printing field values like this:
+       foo: [
+         bar
+       ]
+     rather than something else like
+       foo:
+         [
+           bar
+         ]
+  *)
+let rec format ~inside_box out (x : t) =
```

Comment:
Wasn't it possible to use the yojson's formatters? (I understand that the frontend can't use yojson, but then it also doesn't need to pretty-print the jsons anyway?)

Reply:
The frontend only directly uses `json.ml` to refer to the type `t`, but it also uses some functions in `schema.ml` which in turn call functions in `json.ml`, including some that can error out, with the `error` function (which uses the pretty-printers)

I don't know how to solve this cleanly, though I agree that copy-pasting Yojson code isn't the best

Reply:
One possible option would be to ugly_print in cb-schema, and to pretty_print in cb-check (or even use `jq` as it is even prettier)

Reply:
> One possible option would be to ugly_print in cb-schema, and to pretty_print in cb-check

This is what I ended up doing, and it is much cleaner indeed

Reply:
Ha ok, I now see that extracting informations from a json uses this formatting to provide debugging information on a type error. (My original suggestion was to remove all json pretty-printing code and use `Yojson.Safe.pp..` in place of `Cb_schema.Json.pp` where needed in `cb-check` and the `pipeline`... This could still be done by having the `error` function raise a specific exception catchable by whomever is interested in pretty-printing json, but it's not that important)